src/Symfony/Component/DependencyInjection/Loader/FileLoader.php

```diff
@@ -71,14 +72,17 @@ public function registerClasses(Definition $prototype, $namespace, $resource, $e
                     continue;
                 }
                 foreach (class_implements($class, false) as $interface) {
-                    $singlyImplemented[$interface] = isset($singlyImplemented[$interface]) ? false : $class;
+                    $this->singlyImplemented[$interface] = isset($this->singlyImplemented[$interface]) ? false : $class;
                 }
             }
         }
-        foreach ($interfaces as $interface) {
-            if (!empty($singlyImplemented[$interface])) {
-                $this->container->setAlias($interface, $singlyImplemented[$interface])
-                    ->setPublic(false);
+
+        foreach ($this->interfaces as $interface) {
+            if (!empty($this->singlyImplemented[$interface])) {
+                $this->container->setAlias($interface, $this->singlyImplemented[$interface])->setPublic(false);
+                $this->singlyImplementedAliases[$interface] = true;
+            } elseif ($this->singlyImplementedAliases[$interface] ?? false) {
+                $this->container->removeAlias($interface);
```

Comment:
I think this removes aliases that are defined prior to the call, while it shoudn't. Can you add a test case that does so, by declaring an alias before autodiscovery and ensuring it's still there when several implementation are found by the discovery?
Instead, the implementation should snapshot `getAliases` in the finally block in specific loaders, and restore it at the beginning of this method.

Reply:
The logic/intent here is that _removeAlias_ is called only when an alias was registered by autodiscovery and only then (when alias is registered it is also remembered in _singlyImplementedAliases_).

I thought about a similar approach as you suggested (snapshotting _getAliases_) but I think what I did is enough.

Regardless, I will add another test case for this.

Reply:
In `not_singly_implemented_interface_in_multiple_resources_with_previously_registered_alias`, if you move the definition that loads `resource: ../Prototype/SinglyImplementedInterface/Port/*` just after the alias definition, you will hit the issue I'm mentionning.